### ISSUE TITLE
Remove s as type generic from Canonical

### ISSUE THREAD

**Issue by autholykos** [dusk-network/canonical#75]

**Comment by krl** [comment]
# Changes in the canonical framework.

## The Store trait

The Store was a trait before, implemented by MemStore and DiskStore. These are now combined into a concrete type Store.

Store is a singleton, that has one value across the whole program execution.

There is no need to manually instanciate a MemStore, it's implicitly created.

For persisting to disk, a specific disk-store is created, and given the root of the Canon structure that should be persisted.

## The canon trait

### Before

```rust
/// Trait to read/write values as bytes
pub trait Canon<S: Store>: Sized + Clone {
    /// Write the value as bytes to a `Sink`
    fn write(&self, sink: &mut impl Sink<S>) -> Result<(), S::Error>;
    /// Read the value from bytes in a `Source`
    fn read(source: &mut impl Source<S>) -> Result<Self, S::Error>;
    /// Returns the number of bytes needed to encode this value
    fn encoded_len(&self) -> usize;
}
```

### After

```rust
/// The possible errors when fetching/decoding values from a store
#[derive(Debug)]
pub enum CanonError {
    /// The byte sequence is not a valid representation of the type decoded
    InvalidEncoding,
    /// The instance could not be found in storage
    NotFound,
}

/// Trait to read/write values as bytes
pub trait Canon: Sized + Clone {
    /// Write the value as bytes to a `Sink`
    fn write(&self, sink: &mut Sink);
    /// Read the value from bytes in a `Source`
    fn read(source: &mut Source) -> Result<Self, CanonError>;
    /// Returns the number of bytes needed to encode this value
    fn encoded_len(&self) -> usize;
}
```

### Rationale

Many things going on here.

Sink/Source are no longer traits, but concrete types. 

The write cannot fail anymore, since it's either a write to memory, or a host call. We assume, as does rust error handling in general, that the host does not run out of memory.

Instead of having a error type on read that can incorperate any underlying store error, we only accept two error kinds. Either the value is not found in the backend, or the value is of invalid encoding.

## Consequences

Consumers of the library does not have to worry anymore about the `S` generic of `Canon<S>`, or manually passing around references to `MemStores`

**Comment by krl** [comment]
fixed by #87
